bip-0009.mediawiki

```diff
@@ -0,0 +1,167 @@
+
+In addition, BIP 34 made the integer comparison (nVersion >= 2) a consensus rule after its 95% threshold was reached, removing 2<sup>31</sup>+2 values from the set of valid version numbers (all negative numbers, as nVersion is interpreted as a signed integer, as well as 0 and 1). This indicates another downside this approach: every upgrade permanently restricts the set of allowed nVersion field values. This approach was later reused in BIP 66, which further removed nVersion = 2 as valid option. As will be shown further, this is unnecessary. 
+
+==Specification==
+
+===Mechanism===
+
+'''Bit flags'''
+We are permitting several independent soft forks to be deployed in parallel. For each, a bit B is chosen from the set {0,1,2,...,28}, which is not currently in use for any other ongoing soft fork. Miners signal intent to enforce the new rules associated with the proposed soft fork by setting bit 1<sup>B</sup> in nVersion to 1 in their blocks.
+
+'''High bits'''
+The highest 3 bits are set to 001, so the range of actually possible nVersion values is [0x20000000...0x3FFFFFFF], inclusive. This leaves two future upgrades for different mechanisms (top bits 010 and 011), while complying to the constraints set by BIP34 and BIP66. Having more than 29 available bits for parallel soft forks does not add anything anyway, as the (nVersion >= 3) requirement already makes that impossible.
+
+'''States'''
+With every softfork proposal we associate a state BState, which begins
+at ''defined'', and can be ''locked-in'', ''activated'',
+or ''failed''.  Transitions are considered after each
+retarget period.
+
+'''Soft Fork Support'''
+Software which supports the change should begin by setting B in all blocks
+mined until it is resolved.
+
+    if (BState == defined) {
+        SetBInBlock();
+    }
+
+'''Success: Lock-in Threshold'''
+If bit B is set in 1916 (1512 on testnet) or
+more of the 2016 blocks within a retarget period, it is considered
+''locked-in''.  Miners should stop setting bit B.
+
+    if (NextBlockHeight % 2016 == 0) {
+        if (BState == defined && Previous2016BlocksCountB() >= 1916) {
+            BState = locked-in;
+            BActiveHeight = NextBlockHeight + 2016;
+        }
+    }
+
+'''Success: Activation Delay'''
+The consensus rules related to ''locked-in'' soft fork will be enforced in
+the second retarget period; ie. there is a one retarget period in
+which the remaining 5% can upgrade.  At the that activation block and
+after, the bit B may be reused for a different soft fork.
+
+    if (BState == locked-in && NextBlockHeight == BActiveHeight) {
+        BState = activated;
+        ApplyRulesForBFromNextBlock();
+        /* B can be reused, immediately */
+     }
+
+'''Failure: Timeout'''
+A soft fork proposal should include a ''timeout''.  This is measured
+as the beginning of a calendar year as per this table (suggested
+three years from drafting the soft fork proposal):
```

Comment:
฿tcDrak <notifications@github.com> writes:
>> +'''Success: Activation Delay'''
>> +The consensus rules related to ''locked-in'' soft fork will be enforced in
>> +the second retarget period; ie. there is a one retarget period in
>> +which the remaining 5% can upgrade.  At the that activation block and
>> +after, the bit B may be reused for a different soft fork.
>> +
>> +    if (BState == locked-in && NextBlockHeight == BActiveHeight) {
>> +        BState = activated;
>> +        ApplyRulesForBFromNextBlock();
>> +        /* B can be reused, immediately */
>> +     }
>> +
>> +'''Failure: Timeout'''
>> +A soft fork proposal should include a ''timeout''.  This is measured
>> +as the beginning of a calendar year as per this table (suggested
>> +three years from drafting the soft fork proposal):
>
> three years seems a bit long. 2 maybe, but not more imo.

The proposal is to add 3 to the calendar year, thus 2-3 years.
I clarified this as:

 A soft fork proposal should include a ''timeout''.  This is measured as
 the beginning of a calendar year as per this table (suggest adding
 three to the current calendar year when drafting the soft fork
 proposal):

It can take a year to actually get from first drafting of a BIP to
actual deployed code.  For example, the versionbits bip is 9 months old
already, and still doesn't have code.

Looking at BIP66, the code was released 16 February.  It was activated
on 4 July, about 5 months later.

Cheers,
Rusty.